internal/pipe/milestone/milestone.go

```diff
@@ -0,0 +1,92 @@
+package milestone
+
+import (
+	"github.com/apex/log"
+	"github.com/goreleaser/goreleaser/internal/client"
+	"github.com/goreleaser/goreleaser/internal/git"
+	"github.com/goreleaser/goreleaser/internal/pipe"
+	"github.com/goreleaser/goreleaser/internal/tmpl"
+	"github.com/goreleaser/goreleaser/pkg/config"
+	"github.com/goreleaser/goreleaser/pkg/context"
+)
+
+const defaultNameTemplate = "{{ .Tag }}"
+
+// Pipe for milestone.
+type Pipe struct{}
+
+func (Pipe) String() string {
+	return "milestones"
+}
+
+// Default sets the pipe defaults.
+func (Pipe) Default(ctx *context.Context) error {
+	if len(ctx.Config.Milestones) == 0 {
+		ctx.Config.Milestones = append(ctx.Config.Milestones, config.Milestone{})
+	}
+
+	for i := range ctx.Config.Milestones {
+		milestone := &ctx.Config.Milestones[i]
+
+		if milestone.NameTemplate == "" {
+			milestone.NameTemplate = defaultNameTemplate
+		}
+
+		if milestone.Repo.Name == "" {
+			repo, err := git.ExtractRepoFromConfig()
+
+			if err != nil && !ctx.Snapshot {
+				return err
+			}
+
+			milestone.Repo = repo
+		}
+	}
+
+	return nil
+}
+
+// Publish the release.
+func (Pipe) Publish(ctx *context.Context) error {
+	if ctx.SkipPublish {
+		return pipe.ErrSkipPublishEnabled
+	}
+	c, err := client.New(ctx)
+	if err != nil {
+		return err
+	}
+	return doPublish(ctx, c)
+}
+
+func doPublish(ctx *context.Context, vcsClient client.Client) error {
+	for i := range ctx.Config.Milestones {
+		milestone := &ctx.Config.Milestones[i]
+
+		if !milestone.Close {
+			return pipe.Skip("milestone pipe is disabled")
+		}
+
+		name, err := tmpl.New(ctx).Apply(milestone.NameTemplate)
+
+		if err != nil {
+			return err
+		}
+
+		repo := client.Repo{
+			Name:  milestone.Repo.Name,
+			Owner: milestone.Repo.Owner,
+		}
+
+		log.WithField("milestone", name).
+			WithField("repo", repo.String()).
+			Info("closing milestone")
+
+		err = vcsClient.CloseMilestone(ctx, repo, name)
```

Comment:
I wonder if wether should fail the release if a milestone fails to close 🤔

Reply:
In my feature request, I did have this as a potential configuration option:

```yaml
    # Handling for missing milestone when closing the milestone
    #   error: Prevent release if missing
    #   skip: Skip if missing
    #
    # Default: skip
    missing_on_close: error|skip
```

Please let me know if you would prefer that (or something like it, naming could probably be better) should be implemented and I can sneak that in. 😄

Reply:
I think its a good idea, or just skip by default... wdyt?

Reply:
Settled on this in https://github.com/goreleaser/goreleaser/pull/1657/commits/409fd946f5c785f0f1822d2e30a5411dc2e6c3cf:

```yaml
    # Fail release on errors, such as missing milestone on close
    # Default is false
    fail_on_error: true
```

Which will skip errors by default. Please let me know what you think!